File: zenscript-code-model/src/main/java/raylras/zen/model/type/BoolType.java
package raylras.zen.model.type;

import raylras.zen.model.CompilationEnvironment;
import raylras.zen.model.symbol.Operator;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;
import raylras.zen.model.symbol.SymbolProvider;

import java.util.List;

public enum BoolType implements Type, SymbolProvider {

    INSTANCE;

    @Override
    public String getTypeName() {
        return "bool";
    }

    @Override
    public boolean isCastableTo(Type type, CompilationEnvironment env) {
        if (type instanceof StringType) {
            return true;
        }
        return Type.super.isCastableTo(type, env);
    }

    @Override
    public List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .operator(Operator.NOT, this)
                .operator(Operator.AND, this, params -> params.parameter("value", this))
                .operator(Operator.OR, this, params -> params.parameter("value", this))
                .operator(Operator.XOR, this, params -> params.parameter("value", this))
                .operator(Operator.CONCAT, StringType.INSTANCE, params -> params.parameter("str", StringType.INSTANCE))
                .operator(Operator.EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.NOT_EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.LESS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.LESS_EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.GREATER, BoolType.INSTANCE, params -> params.parameter("value", this))
                .operator(Operator.GREATER_EQUALS, BoolType.INSTANCE, params -> params.parameter("value", this))
                .build();
    }

    @Override
    public String toString() {
        return getTypeName();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/BoolType.kt
package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.Operator
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.SymbolProvider
import raylras.zen.model.symbol.symbolSequence

object BoolType : Type, SymbolProvider {
    override val typeName = "bool"

    override fun isCastableTo(that: Type?, env: CompilationEnvironment): Boolean {
        return when (that) {
            is StringType -> true
            else -> super.isCastableTo(that, env)
        }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        val builtin = symbolSequence {
            operator(Operator.NOT, BoolType)
            operator(Operator.BITWISE_AND, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.BITWISE_OR, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.LOGICAL_AND, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.LOGICAL_OR, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.XOR, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.CONCAT, StringType) {
                parameter("value", StringType)
            }
            operator(Operator.EQUALS, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.NOT_EQUALS, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.LESS, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.LESS_EQUALS, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.GREATER, BoolType) {
                parameter("value", BoolType)
            }
            operator(Operator.GREATER_EQUALS, BoolType) {
                parameter("value", BoolType)
            }
        }
        val classDeclared = env?.classes
            ?.filter { it.qualifiedName == typeName }
            ?.flatMap { it.getSymbols() }
            ?: emptySequence()
        return builtin + classDeclared + getExpands(env)
    }

    override fun toString() = typeName
}
